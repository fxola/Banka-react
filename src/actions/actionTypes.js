export const SIGN_UP_USER = 'SIGN_UP_USER';