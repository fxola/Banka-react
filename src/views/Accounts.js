import React, { Component } from 'react';
import HamburgerIcon from '../components/HamburgerIcon';
import AccountCard from '../components/AccountCard';
import { Link } from 'react-router-dom';
import Footer from '../components/Footer';
import makeRequest from '../services/apiRequest';

class Accounts extends Component {
  state = {
    userAccounts: [],
    isLoading: true
  };

  handleView = () => {};
  handleDelete = () => {};

  async componentDidMount() {
    const token = localStorage.getItem('token');
    const options = {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`
      }
    };
    const response = await makeRequest('/accounts', options);
    this.setState({ userAccounts: response.data, isLoading: false });
  }
  render() {
    const accounts = this.state.userAccounts.map(account => {
      const result = (
        <AccountCard
          key={account.ownerEmail}
          fullName={account.fullName}
          ownerEmail={account.ownerEmail}
          status={account.status}
          balance={account.balance}
          accountNumber={account.accountNumber}
          handleDelete={this.handleDelete}
          handleView={this.handleView}
        />
      );
      return result;
    });

    const template = this.state.isLoading ? <mark>Loading...</mark> : accounts;
    return (
      <>
        <section>
          <header>
            <nav>
              <HamburgerIcon />
              <article className="logo">
                {' '}
                <Link to="/">Banka</Link>
              </article>
              <article className="menu">
                <ul>
                  <Link to="#">
                    <li>Accounts</li>
                  </Link>
                  <Link to="/">
                    <li>Log Out</li>
                  </Link>
                </ul>
              </article>
            </nav>
          </header>
          <main id="admin-dashboard-main">
            <section id="admin-dashboard-section">{template}</section>
          </main>
          <Footer />
        </section>
      </>
    );
  }
}

export default Accounts;
