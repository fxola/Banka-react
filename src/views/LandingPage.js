import React from 'react';
import { Link } from 'react-router-dom';
import HamburgerIcon from '../components/HamburgerIcon';
import Footer from '../components/Footer';

const LandingPage = () => {
  return (
    <section>
      <header>
        <nav id="landing-nav">
          <HamburgerIcon />
          <article className="logo">
            {' '}
            <Link to="/">Banka</Link>
          </article>

          <article className="menu">
            <ul>
              <Link to="/signin">
                <li>Log In</li>
              </Link>
              <Link to="/signup">
                <li>Register</li>
              </Link>
            </ul>
          </article>
        </nav>
        <section id="hero">
          <h1>Want To Experience 21st Century Banking ?</h1>
          <Link to="/signin">
            <button>Let's Go</button>
          </Link>
        </section>
      </header>
      <main id="landing-page-main" />
      <Footer />
    </section>
  );
};

export default LandingPage;
