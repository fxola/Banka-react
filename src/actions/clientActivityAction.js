import * as types from './actionTypes';
import makeRequest from '../services/apiRequest';
export const createAccount = payload => {
  return { type: types.CREATE_BANK_ACCOUNT, payload };
};

export const createAccountError = payload => {
  return { type: types.CREATE_BANK_ACCOUNT_ERROR, payload };
};

export const fetchClientAccountStart = () => {
  return { type: types.FETCH_BANK_ACCOUNT_START };
};
export const fetchClientAccount = payload => {
  return { type: types.FETCH_BANK_ACCOUNT, payload };
};

export const fetchClientAccountError = payload => {
  return { type: types.FETCH_BANK_ACCOUNT_ERROR, payload };
};

export const createAccountRequest = (type, token) => {
  return async dispatch => {
    try {
      const options = {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`
        },
        body: JSON.stringify({ type })
      };
      const response = await makeRequest(`/accounts`, options);
      if (response.status !== 201) throw response;
      dispatch(createAccount(response));
    } catch (error) {
      dispatch(createAccountError(error.message));
    }
  };
};

export const fetchClientAccountRequest = (email, token) => {
  return async dispatch => {
    try {
      dispatch(fetchClientAccountStart());
      const options = {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`
        }
      };
      const response = await makeRequest(`/user/${email}/accounts`, options);
      if (response.status !== 200) throw response;
      dispatch(fetchClientAccount(response.data[0]));
    } catch (error) {
      dispatch(fetchClientAccountError(error.message));
    }
  };
};
