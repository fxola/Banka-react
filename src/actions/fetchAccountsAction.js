import * as types from './actionTypes';
import makeRequest from '../services/apiRequest';
import { checkAuth } from '../services/checkAuth';
export const fetchAccounts = userAccounts => {
  return { type: types.FETCH_ALL_ACCOUNTS, payload: userAccounts };
};

export const fetchAccountsError = error => {
  return { type: types.FETCH_ALL_ACCOUNTS_ERROR, payload: error };
};

export const fetchAccountDetails = accountDetails => {
  return { type: types.FETCH_USER_ACCOUNT_DETAILS, payload: accountDetails };
};

export const fetchAccountDetailsStart = () => {
  return { type: types.FETCH_USER_ACCOUNT_DETAILS_START };
};

export const makeTransaction = payload => {
  return { type: types.MAKE_TRANSACTION, payload };
};
export const cleanUpTransaction = () => {
  return { type: types.CLEAN_UP_TRANSACTION };
};
export const makeTransactionError = payload => {
  return { type: types.MAKE_TRANSACTION_ERROR, payload };
};

export const fetchAccountsRequest = token => {
  return async dispatch => {
    try {
      const options = {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`
        }
      };
      const response = await makeRequest('/accounts', options);
      if (response.status !== 200) throw response;
      dispatch(fetchAccounts(response.data));
    } catch (error) {
      dispatch(fetchAccountsError(error.message));
    }
  };
};

export const fetchAccountDetailsRequest = accountNumber => {
  return async dispatch => {
    try {
      dispatch(fetchAccountDetailsStart());
      const user = JSON.parse(localStorage.getItem('user'));
      const options = {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${user.token}`
        }
      };
      const response = await makeRequest(`/accounts/${accountNumber}`, options);
      // console.log(response);
      if (response.status !== 200) throw response;
      dispatch(fetchAccountDetails(response.data));
    } catch (error) {
      dispatch(fetchAccountsError(error.message));
    }
  };
};

export const makeTransactionRequest = (payload, accountNumber) => {
  return async dispatch => {
    const user = JSON.parse(localStorage.getItem('user'));
    try {
      const options = {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${user.token}`
        },
        body: JSON.stringify({ amount: payload.amount })
      };
      const response = await makeRequest(
        `/transactions/${accountNumber}/${payload.type}`,
        options
      );
      console.log(response);
      if (response.status !== 201) throw response;
      dispatch(makeTransaction(response));
      dispatch(cleanUpTransaction());
    } catch (error) {
      dispatch(makeTransactionError(error.message));
      dispatch(cleanUpTransaction());
    }
  };
};
