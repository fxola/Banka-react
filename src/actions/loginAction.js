import * as types from './actionTypes';
import makeRequest from '../services/apiRequest';
import { toast } from 'react-toastify';
export const logInUser = user => {
  return { type: types.LOG_IN_USER, payload: user };
};

export const loginUserError = error => {
  return { type: types.LOG_IN_USER_ERROR, payload: error };
};

export const logInUserRequest = (user, history) => {
  return async dispatch => {
    try {
      const option = {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(user)
      };
      const response = await makeRequest('/auth/signin', option);

      if (!response.success) {
        if (response.error === 'Network Error') throw response;
        toast.error(response.error);
        dispatch(loginUserError(response.error));
      }

      if (response.success) {
        toast.success('Log in Successful');
        localStorage.setItem('user', JSON.stringify(response.data));
        setTimeout(() => {
          /* istanbul ignore next */
          switch (response.data.type) {
            case 'client':
              history.push('/profile');
              break;
            case 'staff':
              history.push('/accounts');
          }
        }, 3000);
        dispatch(logInUser(response.data));
      }
    } catch (error) {
      toast.error(
        'Something went wrong. Please check your Connection and Try Again'
      );
      dispatch(loginUserError(error.message));
    }
  };
};
