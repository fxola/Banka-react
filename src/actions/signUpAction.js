import * as types from './actionTypes';
import makeRequest from '../services/apiRequest';
import { toast } from 'react-toastify';
export const signUpUser = user => {
  return { type: types.SIGN_UP_USER, payload: user };
};

export const signUpUserError = error => {
  return { type: types.SIGN_UP_USER_ERROR, payload: error };
};

export const signUpUserRequest = (user, history) => {
  return async dispatch => {
    try {
      const option = {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(user)
      };
      const response = await makeRequest('/auth/signup', option);
      if (!response.success) {
        if (response.error === 'Network Error') throw response;
        response.status == 409
          ? toast.error(
              'Email Already Exists. Please provide another email address'
            )
          : toast.error(response.message);
        dispatch(signUpUserError(response.message));
      }
      if (response.success) {
        toast.success(response.message);
        history.push('/profile');
        dispatch(signUpUser(response.data));
      }
    } catch (error) {
      toast.error('Something went wrong. Please try again');
      dispatch(signUpUserError(error.message));
    }
  };
};
