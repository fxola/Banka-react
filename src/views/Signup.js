import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import HamburgerIcon from '../components/HamburgerIcon';
import { signUpUserRequest } from '../actions/signUpAction';
import PropTypes from 'proptypes';

class Signup extends Component {
  state = {
    users: {},
    isloading: false
  };
  firstnameRef = React.createRef();
  lastnameRef = React.createRef();
  emailRef = React.createRef();
  passowordRef = React.createRef();
  confirmPasswordRef = React.createRef();

  createNewuser = async e => {
    e.preventDefault();
    this.setState({ isloading: true });
    const user = {
      firstName: this.firstnameRef.current.value,
      lastName: this.lastnameRef.current.value,
      email: this.emailRef.current.value,
      password: this.passowordRef.current.value,
      confirmPassword: this.confirmPasswordRef.current.value
    };
    try {
      await this.props.dispatch(signUpUserRequest(user));
      this.setState({ isloading: false });
    } catch (error) {
      throw error;
    }
  };

  render() {
    const buttonText = this.state.isloading ? 'Loading...' : 'Register';
    return (
      <section>
        <header>
          <nav>
            <HamburgerIcon />
            <article className="logo">
              {' '}
              <Link to="/">Banka</Link>
            </article>

            <article className="menu">
              <ul>
                <Link to="/signin">
                  <li>Log In</li>
                </Link>
                <Link to="/signup">
                  <li>Register</li>
                </Link>
              </ul>
            </article>
          </nav>
        </header>
        <main id="signup-main">
          <section id="signup-section">
            <form id="signup-form" onSubmit={this.createNewuser}>
              <article id="alert" />
              <input
                type="text"
                name="firstName"
                id="firstname"
                ref={this.firstnameRef}
                placeholder="First Name"
                required
              />
              <input
                type="text"
                name="lastName"
                id="lastname"
                ref={this.lastnameRef}
                placeholder="Last Name"
                required
              />
              <input
                type="email"
                name="email"
                id="email"
                ref={this.emailRef}
                placeholder="Email Address"
                required
              />

              <input
                type="password"
                placeholder="Password"
                name="password"
                ref={this.passowordRef}
                id="password"
                required
              />
              <input
                type="password"
                placeholder="Confirm Password"
                name="confirmPassword"
                ref={this.confirmPasswordRef}
                id="confirmpassword"
                required
              />

              <button id="register-button">{buttonText}</button>
              <p>
                <Link to="#">Forgot Password ?</Link>
              </p>
              <h3>
                Have an account?
                <span className="signup-link">
                  <Link to="/signin">Sign In</Link>
                </span>
              </h3>
            </form>
          </section>
        </main>
        <footer>Hand-crafted By Afolabi &#x00A9; 2019</footer>
      </section>
    );
  }
}

Signup.propTypes = {
  dispatch: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  return {
    users: state.users
  };
};

export default connect(mapStateToProps)(Signup);
