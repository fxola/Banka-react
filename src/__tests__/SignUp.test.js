import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { signUpUserRequest, signUpUserError } from '../actions/signUpAction';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Signup, mapStateToProps } from '../views/Signup';
import { SIGN_UP_USER_ERROR, SIGN_UP_USER } from '../actions/actionTypes';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const renderSignUpPage = args => {
  const defaultProps = {
    isLoading: false,
    history: {},
    users: {},
    match: {},
    dispatch: jest.fn()
  };
  const props = { ...defaultProps, ...args };
  return mount(
    <BrowserRouter>
      <Signup {...props} />
    </BrowserRouter>
  );
};

describe('Sign up Components Tests', () => {
  it('renders Sign Up template', () => {
    const wrapper = renderSignUpPage();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('form').length).toBe(1);
  });

  it('Simulates a form submit event', () => {
    const wrapper = renderSignUpPage();
    wrapper.find('form').simulate('submit');
  });

  it('should show initial state', () => {
    const initialState = {
      users: {}
    };
    expect(mapStateToProps(initialState).users).toEqual({});
  });
});

describe('Sign Up Actions Tests', () => {
  afterEach(() => {
    store.clearActions();
    global.fetch.mockClear();
    delete global.fetch;
  });
  const store = mockStore();
  const userDetails = {
    firstName: 'some',
    lastName: 'user',
    email: 'someuser@gmail.com',
    password: 'randompassword'
  };
  it('Should Trigger the SIGN_UP_USER dispatch function', async () => {
    const mockData = {
      success: true,
      data: {
        firstName: 'some',
        lastName: 'user',
        email: 'someuser@gmail.com',
        token:
          'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MjMsIâ€¦c4OX0.DfJClt28K3qIgJJWp9FMypDO0tKD1NoiNtO9YvEOvsQ'
      }
    };
    const mockJsonPromise = Promise.resolve(mockData);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    const expectedActions = [{ type: SIGN_UP_USER, payload: mockData.data }];
    const historyObject = {
      push: jest.fn()
    };
    await store.dispatch(signUpUserRequest(userDetails, historyObject));
    expect(store.getActions()).toEqual(expectedActions);
  });

  it('Should Trigger the SIGN_UP_USER_ERROR dispatch function on network error', async () => {
    const mockData = {
      success: false,
      error: 'Network Error',
      message: 'Failed To Fetch'
    };
    const mockJsonPromise = Promise.resolve(mockData);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    const historyObject = {
      push: jest.fn()
    };
    const expectedActions = [
      { type: SIGN_UP_USER_ERROR, payload: mockData.message }
    ];
    await store.dispatch(signUpUserRequest(userDetails, historyObject));
    expect(store.getActions()).toEqual(expectedActions);
  });

  it('Should Trigger the LOG_IN_USER_ERROR dispatch function on API error', async () => {
    const mockData = {
      success: false,
      status: 409,
      message: 'Email Already exists'
    };
    const mockJsonPromise = Promise.resolve(mockData);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    const expectedActions = [
      { type: SIGN_UP_USER_ERROR, payload: mockData.message }
    ];
    const historyObject = {
      push: jest.fn()
    };
    await store.dispatch(signUpUserRequest(userDetails, historyObject));
    expect(store.getActions()).toEqual(expectedActions);
  });
});
