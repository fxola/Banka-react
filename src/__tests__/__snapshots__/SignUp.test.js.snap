// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign up Components Tests renders Sign Up template 1`] = `ReactWrapper {}`;
